File: src/f1_final_project/scripts/pure_pursuit_node.py
#!/usr/bin/env python3
import rclpy, os, csv, yaml
import numpy as np
from rclpy.node import Node
from scipy.interpolate import splprep, splev
from nav_msgs.msg import Odometry, Path
from geometry_msgs.msg import PoseStamped, PointStamped
from ackermann_msgs.msg import AckermannDriveStamped
from rcl_interfaces.msg import SetParametersResult
from std_msgs.msg import Float32MultiArray, String

class PurePursuit(Node):
    def __init__(self):
        super().__init__('pure_pursuit')
        
        # --- PARAMETERS ---
        self.declare_parameters(namespace='', parameters=[
            ('waypoint_file', 'ice_waypoints.csv'), 
            ('spacing', 0.15), ('spline_smooth', 0.05),
            ('max_v', 1.5), ('min_v', 0.4), 
            ('max_lat_accel', 0.8), ('max_lon_accel', 0.5), 
            ('lookahead_min', 0.7), ('lookahead_max', 2.5), ('lookahead_gain', 1.5), 
            ('k_sensitivity', 0.5), ('k_lookahead_window', 14),
            ('turn_p', 1.0), ('friction_level', 0) # 0=Use ros params, 1=low, 2=medium, 3=high 
        ])

        self.waypoints = [] # Array of [x, y, velocity, curvature]
        self.pts = [] # Raw waypoints from CSV
        
        # Load friction profiles from YAML
        self.friction_profiles = self.load_friction_profiles()
        
        # --- ROS INFRASTRUCTURE ---
        self.drive_pub = self.create_publisher(AckermannDriveStamped, '/drive', 10)
        self.target_pub = self.create_publisher(PointStamped, '/debug/target_wp', 10)
        self.path_pub = self.create_publisher(Path, '/debug/waypoints_path', 10)
        self.debug_pub = self.create_publisher(Float32MultiArray, '/debug/pure_pursuit_stats', 10)
        self.pose_sub = self.create_subscription(Odometry, '/pf/pose/odom', self.odom_callback, 10)
        self.friction_sub = self.create_subscription(String, '/surface/friction', self.friction_callback, 10)
        
        self.add_on_set_parameters_callback(self.param_callback)
        self.load_and_process()
        self.create_timer(2.0, self.publish_path) # Static path viz

    def friction_callback(self, msg):
        """Updates friction level based on incoming surface condition messages."""
        level_map = {'low': 1, 'medium': 2, 'high': 3}
        new_level = level_map.get(msg.data.lower(), 0)
        
        if new_level != self.get_parameter('friction_level').value:
            self.get_logger().info(f"Friction level changed to: {msg.data} (Level {new_level})")
            self.set_parameters([rclpy.parameter.Parameter('friction_level', rclpy.Parameter.Type.INTEGER, new_level)])

    def load_friction_profiles(self):
        """Loads the friction profiles from the config directory."""
        pkg_path = os.path.join(os.environ.get('AMENT_PREFIX_PATH').split(':')[0], '..', '..', 
                                'src', 'f1_final_project')
        yaml_path = os.path.join(pkg_path, 'config', 'friction_profiles.yaml')
        
        if os.path.exists(yaml_path):
            with open(yaml_path, 'r') as f:
                return yaml.safe_load(f) or {}
        else:
            self.get_logger().warn(f"Friction profiles YAML not found at: {yaml_path}")
            return {}

    def p(self, name):
        """
        Parameter wrapper: Returns YAML value if friction_level > 0, 
        otherwise returns standard ROS parameter value.
        """
        level = self.get_parameter('friction_level').value
        
        # If level is 0, or we failed to load the YAML, use standard ROS param
        if level == 0 or not self.friction_profiles:
            return self.get_parameter(name).value
        
        profile_key = f'level_{level}'
        
        # If the parameter exists in the specific level override, use it.
        # Otherwise, fall back to the ROS parameter.
        if profile_key in self.friction_profiles and name in self.friction_profiles[profile_key]:
            return self.friction_profiles[profile_key][name]
            
        return self.get_parameter(name).value

    def load_and_process(self):
        """Pipeline: Load CSV -> Interpolate -> Spline -> Physics Profile."""
        pkg_path = os.path.join(os.environ.get('AMENT_PREFIX_PATH').split(':')[0], '..', '..', 
                                'src', 'f1_final_project', 'waypoints')
        file_path = os.path.join(pkg_path, self.p('waypoint_file'))
        
        if not os.path.exists(file_path):
            self.get_logger().error(f"File not found: {file_path}")
            return

        with open(file_path, 'r') as f:
            self.pts = np.array([list(map(float, r)) for r in list(csv.reader(f))[1:] if r])
        # 1. GEOMETRY: Linear density followed by spline smoothing
        spacing = self.p('spacing')
        lin_pts = []
        for i in range(len(self.pts)):
            p0, p1 = self.pts[i], self.pts[(i+1) % len(self.pts)]
            vec, dist = p1 - p0, np.linalg.norm(p1 - p0)
            if dist < 0.01: continue
            for j in range(max(int(dist/spacing), 1)): 
                lin_pts.append(p0 + vec * (j/max(int(dist/spacing), 1)))
        
        try:
            tck, _ = splprep(np.array(lin_pts).T, s=self.p('spline_smooth'), per=True)
            smooth_pts = np.array(splev(np.linspace(0, 1, len(lin_pts)), tck)).T
        except: smooth_pts = np.array(lin_pts)

        # STAGE 2: PHYSICS (Circular Curvature Calculation)
        n = len(smooth_pts)
        k, v = np.zeros(n), np.zeros(n)
        win = 3 

        for i in range(n):
            idx_p = (i - win) % n
            idx_n = (i + win) % n
            
            p0, p1, p2 = smooth_pts[idx_p], smooth_pts[i], smooth_pts[idx_n]
            va, vb = p1 - p0, p2 - p1
            la, lb = np.linalg.norm(va), np.linalg.norm(vb)
            
            k[i] = np.arccos(np.clip(np.dot(va,vb)/(la*lb + 1e-6), -1, 1)) / ((la+lb)/2 + 1e-6)
            v[i] = np.sqrt(self.p('max_lat_accel') / (k[i] + 1e-6))
        
        v = np.clip(v, self.p('min_v'), self.p('max_v'))
        
        # Forward/Backward passes to respect longitudinal acceleration limits
        alon = self.p('max_lon_accel')
        for _ in range(2):
            for i in reversed(range(n)):
                nxt = (i+1)%n
                ds = np.linalg.norm(smooth_pts[nxt]-smooth_pts[i])
                v[i] = min(v[i], np.sqrt(v[nxt]**2 + 2*alon*ds))
            for i in range(n):
                prv = (i-1)%n
                ds = np.linalg.norm(smooth_pts[i]-smooth_pts[prv])
                v[i] = min(v[i], np.sqrt(v[prv]**2 + 2*alon*ds))

        self.waypoints = np.column_stack((smooth_pts, v, k))

    def odom_callback(self, msg):
        if not len(self.waypoints): return
        
        curr = np.array([msg.pose.pose.position.x, msg.pose.pose.position.y])
        idx = np.argmin(np.linalg.norm(self.waypoints[:,:2] - curr, axis=1))
        
        # 3. ADAPTIVE LOOKAHEAD: Shrink Ld if high curvature is detected ahead
        win = self.p('k_lookahead_window')
        ahead_k = np.max(self.waypoints[[(idx + j - 2) % len(self.waypoints) for j in range(win)], 3])
        
        Ld = (self.p('lookahead_gain') * self.waypoints[idx, 2]) / \
             (1.0 + (ahead_k * self.p('k_sensitivity')))
        Ld = np.clip(Ld, self.p('lookahead_min'), self.p('lookahead_max'))

        # 4. PURSUIT: Find target point and calculate steering
        target = self.waypoints[idx]
        for i in range(len(self.waypoints)):
            wp = self.waypoints[(idx + i) % len(self.waypoints)]
            if np.linalg.norm(wp[:2] - curr) > Ld:
                target = wp
                break

        # Transform target to local frame for lateral error (ly)
        q = msg.pose.pose.orientation
        yaw = np.arctan2(2*(q.w*q.z + q.x*q.y), 1-2*(q.y*q.y + q.z*q.z))
        dx, dy = target[0]-curr[0], target[1]-curr[1]
        ly = -dx * np.sin(yaw) + dy * np.cos(yaw) 
        
        # Command Drive
        drive = AckermannDriveStamped()
        drive.header.stamp = self.get_clock().now().to_msg()
        drive.drive.speed = float(target[2])
        drive.drive.steering_angle = np.clip(float((self.p('turn_p') * ly) / (Ld**2)), -0.4, 0.4)
        self.drive_pub.publish(drive)

        # Debug Visuals
        tp = PointStamped()
        tp.header.frame_id, tp.header.stamp = "map", self.get_clock().now().to_msg()
        tp.point.x, tp.point.y = target[0], target[1]
        self.target_pub.publish(tp)

        # Create debug message
        debug_msg = Float32MultiArray()
        debug_msg.data = [float(Ld), float(target[2]), float(ahead_k), float(ly)]
        self.debug_pub.publish(debug_msg)

    def param_callback(self, params):
        # Deferred execution via one-shot timer ensures params are fully saved in ROS map
        self.update_timer = self.create_timer(0.1, self.delayed_load)
        return SetParametersResult(successful=True)

    def delayed_load(self):
        self.update_timer.cancel()
        self.destroy_timer(self.update_timer)
        self.load_and_process()

    def publish_path(self):
        if not len(self.waypoints): return
        msg = Path()
        msg.header.frame_id, msg.header.stamp = "map", self.get_clock().now().to_msg()
        for wp in self.waypoints:
            p = PoseStamped()
            p.pose.position.x, p.pose.position.y = wp[0], wp[1]
            msg.poses.append(p)
        self.path_pub.publish(msg)

def main():
    rclpy.init(); rclpy.spin(PurePursuit()); rclpy.shutdown()

if __name__ == '__main__': main()

File: src/f1_final_project/src/pure_pursuit_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <nav_msgs/msg/path.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <geometry_msgs/msg/point_stamped.hpp>
#include <ackermann_msgs/msg/ackermann_drive_stamped.hpp>
#include <std_msgs/msg/float32_multi_array.hpp>
#include <rcl_interfaces/msg/set_parameters_result.hpp>

#include <vector>
#include <string>
#include <fstream>
#include <sstream>
#include <cmath>
#include <algorithm>

struct WP {
    double x, y, v, k;
};

class PurePursuit : public rclcpp::Node {
public:
    PurePursuit() : Node("pure_pursuit") {
        // --- PARAMETERS ---
        this->declare_parameter("waypoint_file", "my_map_wp.csv");
        this->declare_parameter("spacing", 0.15);
        this->declare_parameter("spline_smooth", 0.05);
        this->declare_parameter("max_v", 3.0);
        this->declare_parameter("min_v", 0.5);
        this->declare_parameter("max_lat_accel", 1.5);
        this->declare_parameter("max_lon_accel", 1.0);
        this->declare_parameter("lookahead_min", 0.7);
        this->declare_parameter("lookahead_max", 2.5);
        this->declare_parameter("lookahead_gain", 1.5);
        this->declare_parameter("k_sensitivity", 1.2);
        this->declare_parameter("k_lookahead_window", 14);
        this->declare_parameter("turn_p", 1.5);

        // --- ROS INFRASTRUCTURE ---
        drive_pub = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>("/drive", 10);
        target_pub = this->create_publisher<geometry_msgs::msg::PointStamped>("/debug/target_wp", 10);
        path_pub = this->create_publisher<nav_msgs::msg::Path>("/debug/waypoints_path", 10);
        debug_pub = this->create_publisher<std_msgs::msg::Float32MultiArray>("/debug/pure_pursuit_stats", 10);
        
        pose_sub = this->create_subscription<nav_msgs::msg::Odometry>(
            "/pf/pose/odom", 10, std::bind(&PurePursuit::odom_callback, this, std::placeholders::_1));

        param_handler = this->add_on_set_parameters_callback(
            std::bind(&PurePursuit::param_callback, this, std::placeholders::_1));

        load_and_process();
        path_timer = this->create_wall_timer(std::chrono::milliseconds(2000), std::bind(&PurePursuit::publish_path, this));
    }

private:
    void load_and_process() {
        std::string filename = this->get_parameter("waypoint_file").as_string();
        std::string file_path = "../../src/f1_final_project/waypoints/" + filename;
        
        std::ifstream file(file_path);
        if (!file.is_open()) {
            RCLCPP_ERROR(this->get_logger(), "File not found: %s", file_path.c_str());
            return;
        }

        std::vector<std::pair<double, double>> pts;
        std::string line;
        std::getline(file, line); 
        while (std::getline(file, line)) {
            if (line.empty()) continue;
            std::stringstream ss(line);
            std::string item;
            std::vector<double> row;
            while (std::getline(ss, item, ',')) row.push_back(std::stod(item));
            if (row.size() >= 2) pts.push_back({row[0], row[1]});
        }

        double spacing = this->get_parameter("spacing").as_double();
        std::vector<std::pair<double, double>> smooth_pts;
        for (size_t i = 0; i < pts.size(); ++i) {
            auto p0 = pts[i];
            auto p1 = pts[(i + 1) % pts.size()];
            double dx = p1.first - p0.first, dy = p1.second - p0.second;
            double dist = std::sqrt(dx*dx + dy*dy);
            if (dist < 0.01) continue;
            int steps = std::max((int)(dist / spacing), 1);
            for (int j = 0; j < steps; ++j) {
                smooth_pts.push_back({p0.first + dx * (j / (double)steps), p0.second + dy * (j / (double)steps)});
            }
        }

        size_t n = smooth_pts.size();
        waypoints.clear();
        waypoints.resize(n);
        int win = 3;
        double max_lat = this->get_parameter("max_lat_accel").as_double();
        double min_v = this->get_parameter("min_v").as_double();
        double max_v = this->get_parameter("max_v").as_double();

        for (size_t i = 0; i < n; ++i) {
            int ip = (i - win + n) % n, in = (i + win) % n;
            auto p0 = smooth_pts[ip], p1 = smooth_pts[i], p2 = smooth_pts[in];
            double vax = p1.first - p0.first, vay = p1.second - p0.second;
            double vbx = p2.first - p1.first, vby = p2.second - p1.second;
            double la = std::sqrt(vax*vax + vay*vay), lb = std::sqrt(vbx*vbx + vby*vby);
            double dot = vax*vbx + vay*vby;
            double k = std::acos(std::clamp(dot / (la * lb + 1e-6), -1.0, 1.0)) / ((la + lb) / 2.0 + 1e-6);
            double v = std::sqrt(max_lat / (k + 1e-6));
            waypoints[i] = {p1.first, p1.second, std::clamp(v, min_v, max_v), k};
        }

        double alon = this->get_parameter("max_lon_accel").as_double();
        for (int pass = 0; pass < 2; ++pass) {
            for (int i = (int)n - 1; i >= 0; --i) {
                int nxt = (i + 1) % n;
                double ds = std::sqrt(std::pow(waypoints[nxt].x - waypoints[i].x, 2) + std::pow(waypoints[nxt].y - waypoints[i].y, 2));
                waypoints[i].v = std::min(waypoints[i].v, std::sqrt(std::pow(waypoints[nxt].v, 2) + 2 * alon * ds));
            }
            for (size_t i = 0; i < n; ++i) {
                int prv = (i - 1 + n) % n;
                double ds = std::sqrt(std::pow(waypoints[i].x - waypoints[prv].x, 2) + std::pow(waypoints[i].y - waypoints[prv].y, 2));
                waypoints[i].v = std::min(waypoints[i].v, std::sqrt(std::pow(waypoints[prv].v, 2) + 2 * alon * ds));
            }
        }
    }

    void odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg) {
        if (waypoints.empty()) return;
        double cx = msg->pose.pose.position.x, cy = msg->pose.pose.position.y;
        size_t idx = 0;
        double min_d = 1e18;
        for (size_t i = 0; i < waypoints.size(); ++i) {
            double d = std::pow(waypoints[i].x - cx, 2) + std::pow(waypoints[i].y - cy, 2);
            if (d < min_d) { min_d = d; idx = i; }
        }

        int win = this->get_parameter("k_lookahead_window").as_int();
        double ahead_k = 0;
        for (int j = 0; j < win; ++j) ahead_k = std::max(ahead_k, waypoints[(idx + j) % waypoints.size()].k);

        double Ld = (this->get_parameter("lookahead_gain").as_double() * waypoints[idx].v) / 
                    (1.0 + (ahead_k * this->get_parameter("k_sensitivity").as_double()));
        Ld = std::clamp(Ld, this->get_parameter("lookahead_min").as_double(), this->get_parameter("lookahead_max").as_double());

        WP target = waypoints[idx];
        for (size_t i = 0; i < waypoints.size(); ++i) {
            const auto& wp = waypoints[(idx + i) % waypoints.size()];
            if (std::sqrt(std::pow(wp.x - cx, 2) + std::pow(wp.y - cy, 2)) > Ld) { target = wp; break; }
        }

        auto q = msg->pose.pose.orientation;
        double yaw = std::atan2(2.0*(q.w*q.z + q.x*q.y), 1.0 - 2.0*(q.y*q.y + q.z*q.z));
        double dx = target.x - cx, dy = target.y - cy;
        double ly = -dx * std::sin(yaw) + dy * std::cos(yaw);

        auto drive = ackermann_msgs::msg::AckermannDriveStamped();
        drive.header.stamp = this->now();
        drive.drive.speed = (float)target.v;
        drive.drive.steering_angle = std::clamp((float)((this->get_parameter("turn_p").as_double() * ly) / (Ld * Ld)), -0.4f, 0.4f);
        drive_pub->publish(drive);

        std_msgs::msg::Float32MultiArray dbg;
        dbg.data = {(float)Ld, (float)target.v, (float)ahead_k, (float)ly};
        debug_pub->publish(dbg);
    }

    rcl_interfaces::msg::SetParametersResult param_callback(const std::vector<rclcpp::Parameter> &params) {
        (void)params;
        this->create_wall_timer(std::chrono::milliseconds(100), [this]() { this->load_and_process(); });
        rcl_interfaces::msg::SetParametersResult result;
        result.successful = true;
        result.reason = "OK";
        return result;
    }

    void publish_path() {
        if (waypoints.empty()) return;
        auto msg = nav_msgs::msg::Path();
        msg.header.frame_id = "map";
        msg.header.stamp = this->now();
        for (const auto& wp : waypoints) {
            geometry_msgs::msg::PoseStamped p;
            p.pose.position.x = wp.x; p.pose.position.y = wp.y;
            msg.poses.push_back(p); // CHANGED FROM append TO push_back
        }
        path_pub->publish(msg);
    }

    std::vector<WP> waypoints;
    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr drive_pub;
    rclcpp::Publisher<geometry_msgs::msg::PointStamped>::SharedPtr target_pub;
    rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr path_pub;
    rclcpp::Publisher<std_msgs::msg::Float32MultiArray>::SharedPtr debug_pub;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr pose_sub;
    rclcpp::TimerBase::SharedPtr path_timer;
    OnSetParametersCallbackHandle::SharedPtr param_handler;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<PurePursuit>());
    rclcpp::shutdown();
    return 0;
}